### ISSUE TITLE
`statsmodels.iolib.summary2.summary_col` bug in `regressor_order`

### ISSUE THREAD

**Issue by jiafengkevinchen** [statsmodels/statsmodels#3767]
![image](https://user-images.githubusercontent.com/24930289/27343801-115d10ae-55b2-11e7-9430-31b483a5c930.png)
regs is a list of `RegressionResults`. The `regressor_order` command seems to change only the regressor names and not permute the coefficients appropriately

**Comment by josef-pkt** [comment]
(issue got lost without labels)

I don't know if this is still the case in master/soon 0.9
needs example and checking

**Comment by bashtage** [comment]
Added a test that shows this isn't an issue.

**Comment by whoopnip** [comment]
Hi, I'm still running into this issue on 0.8.0. Looking at the source code, I was able to track down the problem, why the test did not catch it, and also propose a solution.

The issue only occurs when the regressors are not in alphabetical order (as is the case for the person who opened the issue). For the test, the two regressors are x1 and x2, and const is displayed first, so they are already in alphabetical order.

The relevant code from `summary_col` is:

```
    if regressor_order:
        varnames = summ.index.get_level_values(0).tolist()
        ordered = [x for x in regressor_order if x in varnames]
        unordered = [x for x in varnames if x not in regressor_order + ['']]
        order = ordered + list(np.unique(unordered))

        f = lambda idx: sum([[x + 'coef', x + 'stde'] for x in idx], [])
        summ.index = f(np.unique(varnames))
        summ = summ.reindex(f(order))
        summ.index = [x[:-4] for x in summ.index]
```

The problem comes from using `np.unique`. The line using it is changing the MultiIndex into a single-level index where the index values are the regressor names + "coef" or "stde" to identify coefficients and standard errors. That way in the following line, we can reindex the DataFrame with the correct order. 

That line using `np.unique` is supposed to replace the indices but not change their order. The changing of order is only supposed to be done with the reindex. But `np.unique` sorts during the process, and returns an alphabetically sorted list, so it is changing the order of the index when reassigning to the DataFrame, while the coefficients and standard errors remain in place. This causes a mismatch of the values and the corresponding variables.

There is a very easy fix. We can just go to using `pd.unique` instead of `np.unique`, which does retain order. I have tested it and it drops right in place without any other changes necessary.

**Comment by josef-pkt** [comment]
@whoopnip This should be fixed in master according to the unit test for closing this issue.

Can you provide an example that fails for you so I can check with master?

**Comment by whoopnip** [comment]
@josef-pkt Sure. We can modify the test code to produce the error:

```
import pandas as pd
import numpy as np
from statsmodels.api import add_constant, OLS
from statsmodels.iolib.summary2 import summary_col

x = [1, 5, 7, 3, 5]
x = add_constant(x)
x2 = np.concatenate([x, np.array([[3], [9], [-1], [4], [0]])], 1)
x2 = pd.DataFrame(x2, columns=['const','b','a']) # ensure that columns are not in alphabetical order
y1 = [6, 4, 2, 7, 4]
y2 = [8, 5, 0, 12, 4]
reg1 = OLS(y1, x2).fit()
reg2 = OLS(y2, x2).fit()

info_dict = {'R2': lambda x: '{:.3f}'.format(int(x.rsquared)),
             'N': lambda x: '{0:d}'.format(int(x.nobs))}
original = actual = summary_col([reg1, reg2], float_format='%0.4f')
actual = summary_col([reg1, reg2], regressor_order=['a', 'b'],
                     float_format='%0.4f',
                     info_dict=info_dict)

>>> original
=======================
        y I      y II  
-----------------------
const 7.4428   11.1742 
      (1.5638) (4.3596)
b     -0.7206  -1.4573 
      (0.2926) (0.8157)
a     0.0612   0.2488  
      (0.1695) (0.4725)
=======================
Standard errors in
parentheses.

>>> actual

=======================
        y I      y II  
-----------------------
a     7.4428   11.1742 
      (1.5638) (4.3596)
b     -0.7206  -1.4573 
      (0.2926) (0.8157)
const 0.0612   0.2488  
      (0.1695) (0.4725)
R2    0.000    0.000   
N     5        5       
=======================
Standard errors in
parentheses.
```

You could replace the full test by:

```
def test_summary_col_ordering_preserved(self):
    x = [1, 5, 7, 3, 5]
    x = add_constant(x)
    x2 = np.concatenate([x, np.array([[3], [9], [-1], [4], [0]])], 1)
    x2 = pd.DataFrame(x2, columns=['const','b','a'])
    y1 = [6, 4, 2, 7, 4]
    y2 = [8, 5, 0, 12, 4]
    reg1 = OLS(y1, x2).fit()
    reg2 = OLS(y2, x2).fit()

    info_dict = {'R2': lambda x: '{:.3f}'.format(int(x.rsquared)),
                 'N': lambda x: '{0:d}'.format(int(x.nobs))}
    original = actual = summary_col([reg1, reg2], float_format='%0.4f')
    actual = summary_col([reg1, reg2], regressor_order=['a', 'b'],
                         float_format='%0.4f',
                         info_dict=info_dict)
    variables = ('const', 'b', 'a')
    for line in str(original).split('\n'):
        for variable in variables:
            if line.startswith(variable):
                assert line in str(actual)
```

Let me know if I can help in any other way.

**Comment by josef-pkt** [comment]
Thanks, I get the same wrong result in master

**Comment by josef-pkt** [comment]
@whoopnip  Do you want to submit this as a PR?   
using pandas unique sounds good to me, and making sure there is a testcase that fails without the change and succeeds with it.

If not, then I will make a PR within a few days.

Thanks, both for report and suggested fix

**Comment by josef-pkt** [comment]
fixed in #4271
